### ISSUE TITLE
Run Profile Without a Transaction?

### ISSUE THREAD

**Issue by jzabroski** [fluentmigrator/fluentmigrator#966]
It does not seem possible to specify TransactionBehavior.None on a profile.  The ProfileAttribute and Migration class are not 100% composable for this reason: MigrationAttribute specifies a `long version` identifier, which is persisted to the `dbo.VersionInfo` table.  Profiles do not belong in this table, although it would be nice to log that a profile ran.  In RoundhousE (a competing tool I used to use but now prefer FM), they would call these profiles "Anytime scripts", and they would log these to the VersionInfo table.

MigrationAttribute has the following constructors:
1. Inherit transaction of command line runner:
   1. `public MigrationAttribute(long version)`
   2. `public MigrationAttribute(long version, string description)`
2. Use transaction specified in migration:
   1. `public MigrationAttribute(long version, TransactionBehavior transactionBehavior)`
   2. `public MigrationAttribute(long version, TransactionBehavior transactionBehavior, string description)`

I don't mean to be pedantic spelling all this out for you, Mark, just thought I would highlight the need for an orthogonal way to specify transactions.

**Comment by jzabroski** [comment]
My workaround for this is to create a migration with a negative ID and create a profile that deletes all negative migration numbers.

**Comment by fubar-coder** [comment]
That's a clever workaround, but that would definitely be a useful feature.

**Comment by jzabroski** [comment]
Here is another scenario I have run into where Profiles Without Transactions missing feature has screwed me: I want to run, every time I run migrations, a profile for each environment that ensures (a) linked servers are set-up correctly (b) synonyms are put in place to abstract away the four part name for the linked server.

As it turns out, in reseaching this issue, TransactionBehavior.None on migrations was added to FM many years ago precisely for linked servers: https://www.codesd.com/item/how-to-create-a-linked-server-using-fluentmigrator.html - However, the feature never made it in for Profiles, where it probably makes the most sense, given it's a property of the database's enclosing environment (the server), rather than the database itself. In other words, writing to the VersionInfo table that the script has been run is incorrect, because as the environment changes, the VersionInfo table is unaware of such changes.

**Comment by jzabroski** [comment]
For the add linked server scenario, here is my workaround:

```csharp
        public void AddLinkedServerIfNotExists(string remoteServerName)
        {
            // If this fails, then you may need to grant ALTER ANY LINKED SERVER permission to the account running FluentMigrator.
            // Alternatively, the sysadmin fixed server role has this permission as well.
            Execute.WithConnection((conn, tran) =>
            {
                var cmdText = $@"
IF NOT EXISTS (
    SELECT * FROM sys.servers s WHERE s.is_linked = CAST(1 AS BIT) AND s.[name] = '{remoteServerName}'
)
BEGIN
    EXEC ('sp_addlinkedserver @server = ''{remoteServerName}''');
END
";
                using (var conn2 = new SqlConnection(tran.Connection.ConnectionString))
                {
                    using (var cmd = conn2.CreateCommand())
                    {
                        cmd.CommandText = cmdText;
                        conn2.Open();
                        cmd.ExecuteNonQuery();
                    }
                }
            });
        }
```

**Comment by jzabroski** [comment]
@fubar-coder In upgrading to 3.2, my workaround here for linked servers no longer works.  Thinking about taking you up on your offer to be a co-maintainer.  Do you want to reach out via Gmail and I can share my WhatsApp?

Note: Your Contact link on your website is failing. You are probably losing prospective customers. Inspecting the 200 response shows: `{"into":"#wpcf7-f174-p170-o1","status":"mail_failed","message":"The message couldn't be sent. Please try again at a later time or notify the web site administrator."}`

**Comment by fubar-coder** [comment]
Damn, right now I only have access to a smartphone. The easiest thing would be for you to join the Gitter channel and write me a message. I can give you my contact information there. Or you can give me yours.

**Comment by jzabroski** [comment]
I don't see a gitter badge for FM in the README.md - what is it?

**Comment by fubar-coder** [comment]
It's https://gitter.im/FluentMigrator/fluentmigrator

**Comment by fubar-coder** [comment]
@jzabroski E-Mail is working again.

**Comment by fubar-coder** [comment]
@jzabroski Ok, so I took a look and profiles are indeed **not** what you need to use. You should use "maintenance migrations" instead. They are like migrations, but with a `MaintenanceAttribute` where you're able to configure the transaction behaviour. One with `MigrationStage.BeforeAll` and another with `MigrationStage.AfterAll`.

**Comment by jzabroski** [comment]
Maintenance migrations - I didn't know this existed and this is exactly what RoundhousE used to provide.  Excellent.

Some questions and self-documentation as I go to try this feature out:

- Do MaintenanceAttribute migrations respect Tags? e.g., I have Tags for each stage in a continuous delivery lifecycle: my Tags are UnitTest, Dev, Alpha, Beta, Production.  I want a unique Maintenance migration for each stage.
  - My use case is to create synonyms for database objects, and whether I use linked servers with four part names or local databases with three party names; if a linked server is required, I must create it.
- Do Maintenance migrations run even when there are no regular migrations to run? e.g., if I run migrate up back to back without adding migrations, and the set `All` is empty, what's the behavior?
  - This is more so I just understand how it works.  I can also see if there is code coverage to enforce such concepts.

**Comment by fubar-coder** [comment]
Yes, they redirect tags, but untagged maintenence  classes are ignored by default when tags were specified at the command line. This behavior is configurable.

Yes, they are always run.
